gcloud/datastore/query.py

```diff
@@ -463,6 +463,13 @@ def __iter__(self):
                 yield entity
             if not self._more_results:
                 break
+            num_results = len(self._page)
+            if self._limit is not None:
+                self._limit -= num_results
+            if self._offset is not None:
+                # NOTE: The offset goes down relative to the location
+                #       because we are updating the cursor each time.
+                self._offset -= num_results
```

Comment:
I'm not sure `num_results` is correct here. When datastore satisfies the offset the size of the page isn't related. You need to use the value of `num_skipped_results` from the call to datastore.